src/lavinmq/vhost.cr

```diff
@@ -369,18 +369,68 @@ module LavinMQ
       upstreams.stop_all
     end
 
+    private def close_connections(reason, close_done : Channel(Nil))
+      ch = Channel(Client).new
+      wg = WaitGroup.new
+
+      fiber_count = 0
+      spawn_close_fiber = -> do
+        fiber_id = (fiber_count &+= 1)
+        @log.trace { "Spawn close fiber #{fiber_id}" }
+        wg.add
+        spawn(name: "vhost close connection #{fiber_id}") do
+          loop do
+            select
+            when client = ch.receive
+              client.close(reason)
+            when timeout 50.milliseconds
+              break
+            end
+          rescue
+          end
+          wg.done
+          @log.trace { "Exit close fiber #{fiber_id}" }
+        end
+      end
+
+      {@connections.size, 100}.min.times do
+        spawn_close_fiber.call
+      end
+
+      @connections.each do |client|
+        loop do
+          select
+          when ch.send client
+            break
+          when timeout 200.milliseconds
+            # Spawn more fibers if all current are too busy closing "zombies"
+            spawn_close_fiber.call
+          end
+        end
+      end
+
+      wg.wait
+      ch.close
```

Comment:
close the channel first, so that the fibers can stop, then wait for all fibers to finish

```suggestion
      ch.close
      wg.wait
```